### PR TITLE
perf: make `.dt.truncate('*mo')` more than 3x faster

### LINKED ISSUE #13157: dt.truncate is slow
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import duckdb
import polars as pl
import numpy as np

start = np.datetime64('2015-01-01', 's').astype('int')
end = np.datetime64('2020-01-01', 's').astype('int')
n = 100_000_000

df = pl.DataFrame({'date': np.random.randint(start, end, n).astype('datetime64[s]').astype('datetime64[ms]')})

# looks like only one CPU core is utilised, same for `1y`
%time df.group_by(pl.col.date.dt.truncate('1mo')).agg(pl.count())
# CPU times: user 37 s, sys: 1.15 s, total: 38.1 s
# Wall time: 34.7 s

# though for '1d' it is significantly faster
%time df.group_by(pl.col.date.dt.truncate('1d')).agg(pl.count())
# CPU times: user 7.16 s, sys: 634 ms, total: 7.79 s
# Wall time: 4.38 s

# but duckdb is much faster
%time duckdb.sql("select date_trunc('month', date) as month, count() from df group by month").pl()
# CPU times: user 7.02 s, sys: 93.1 ms, total: 7.12 s
# Wall time: 1.27 s

# even if using time_bucket function which is probably a closer alternative to polars dt.truncate
%time duckdb.sql("select time_bucket(interval '1 month', date) as month, count() from df group by month").pl()
# CPU times: user 11.9 s, sys: 108 ms, total: 12 s
# Wall time: 1.88 s
```

### Log output

_No response_

### Issue description

`dt.truncate` is slow when used with intervals >= '1mo', for smaller intervals it seems to be faster, though still it falls behind duckdb. Also seems like it doesn't utilise all CPU cores.

### Expected behavior

At least `dt.truncate` for month intervals to work as fast as for smaller intervals, or even better if it could match duckdb performance.

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.1
Index type:           UInt32
Platform:             macOS-10.16-x86_64-i386-64bit
Python:               3.10.13 (main, Sep 11 2023, 08:39:02) [Clang 14.0.6 ]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fsspec:               2023.12.1
gevent:               <not installed>
matplotlib:           3.7.1
numpy:                1.24.2
openpyxl:             <not installed>
pandas:               2.1.4
pyarrow:              14.0.1
pydantic:             2.1.1
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>

duckdb:               0.9.2
```

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #13157 

benchmark: https://www.kaggle.com/code/marcogorelli/polars-timing?scriptVersionId=155987273

going to take my time on this one, as there's a few details which are easy to get wrong and I don't think the current test coverage covers everything...looks promising though

### CODE REVIEW

File: crates/polars-time/src/windows/duration.rs [hunk 1]
```diff
@@ -544,48 +543,68 @@ impl Duration {
...
+        let mut result = (original_dt_local.day() - 1) as i64;
+        while remainder > 0 {
+            month -= 1;
+            if month == 0 {
+                year -= 1;
+            }
+            else if month < 0 {
+                month = 11;
+            }
+            let days = if is_leap_year(year as i32) {
```
[the remark this answers is not in the document]
Comment (ritchie46) [reply crates/polars-time/src/windows/duration.rs]: And any improvement by this?

[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply crates/polars-time/src/windows/duration.rs]: nothing too perceptible by the looks of it...I think the biggest gain is just having avoided one "create datetime then get timestamp" cycle

it does clean things up though, thanks 🙏

File: py-polars/tests/parametric/time_series/test_truncate.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    n=st.integers(min_value=1, max_value=100),
+)
+def test_truncate_monthly(value: dt.date, n: int) -> None:
+    result = pl.Series([value]).dt.truncate(f"{n}mo").item()
+    # manual calculation
+    total = value.year * 12 + value.month - 1
+    remainder = total % n
+    total -= remainder
+    year, month = (total // 12), ((total % 12) + 1)
+    expected = dt.datetime(year, month, 1)
```
Comment (MarcoGorelli) [inline py-polars/tests/parametric/time_series/test_truncate.py:22]: this is essentially the old logic, so just checking there's no breaking changes here

File: crates/polars-time/src/windows/duration.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ritchie46) [reply crates/polars-time/src/windows/duration.rs]: Alright. Nice to have 3x here. Would love a fast chrono. :')

**Comment by ritchie46** [comment]
Great one @MarcoGorelli. Is it ready?

**Comment by MarcoGorelli** [comment]
thanks! yeah if you're happy with it, I think it should be robust (at least, the new parametric test is extensive and helped find some edge cases along the way, and the rest of the test suite is still green)
